provider/pkg/internal/pulumiapi/template_source.go

```diff
@@ -0,0 +1,82 @@
+package pulumiapi
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"path"
+)
+
+type CreateTemplateSourceRequestDestination struct {
+	URL *string `json:"url"`
+}
+
+type CreateTemplateSourceRequest struct {
+	Name        string                                  `json:"name"`
+	SourceURL   string                                  `json:"sourceURL"`
+	Destination *CreateTemplateSourceRequestDestination `json:"destination"`
+}
+
+type TemplateSourceResponse struct {
+	Id          string                                  `json:"id"`
+	IsValid     bool                                    `json:"isValid"`
+	Name        string                                  `json:"name"`
+	SourceURL   string                                  `json:"sourceURL"`
+	Destination *CreateTemplateSourceRequestDestination `json:"destination"`
+}
+
+type ListResponse struct {
+	Sources []TemplateSourceResponse `json:"sources"`
+}
+
+func (c *Client) CreateTemplateSource(ctx context.Context, organizationName string, request CreateTemplateSourceRequest) (*TemplateSourceResponse, error) {
+	apiPath := path.Join("orgs", organizationName, "templates/sources")
+	var response TemplateSourceResponse
+	_, err := c.do(ctx, http.MethodPost, apiPath, request, &response)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create template source in org %s: %+v due to error: %w", organizationName, request, err)
+	}
+	return &response, nil
+}
+
+func (c *Client) UpdateTemplateSource(ctx context.Context, organizationName string, templateId string, request CreateTemplateSourceRequest) (*TemplateSourceResponse, error) {
+	apiPath := path.Join("orgs", organizationName, "templates/sources", templateId)
+	var response TemplateSourceResponse
+	_, err := c.do(ctx, http.MethodPatch, apiPath, request, &response)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update template source in org %s with id %s: %+v due to error: %w", organizationName, templateId, request, err)
+	}
+	return &response, nil
+}
+
+func (c *Client) GetTemplateSource(ctx context.Context, organizationName string, templateID string) (*TemplateSourceResponse, error) {
+	// This sucks, but there's not Get API for Template Sources
+	// Thus, using a List and then finding by ID
+
+	apiPath := path.Join("orgs", organizationName, "templates/sources")
+	var templateSources ListResponse
+	_, err := c.do(ctx, http.MethodGet, apiPath, nil, &templateSources)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get template source in org %s with id %s due to error: %w", organizationName, templateID, err)
+	}
+
+	for _, source := range templateSources.Sources {
+		if source.Id == templateID {
+			return &source, nil
+		}
+	}
+
+	return nil, nil
```

Comment:
not entirely sure of the pattern here but if we can't find a template source do we want to raise an error?

Reply:
We don't want to throw an error here, for example during refresh we expect it to be not found (if user deleted the template source, causing a drift)

I mean, I could refactor this to return NotFound error here instead and have caller handle that, but imho this way is better, since there was no error calling API, just nothing (nil) found